### PR TITLE
Plot data card - for making custom graph

### PR DESCRIPTION
**Pull request by Danielhiversen**
This will add a possibility for a component to make a custom graph. All components can provide a _plot_data dictionary as an attribute, and this will result in a custom graph in the frontend.

untitled

I am planing to use it in the backend like this: https://github.com/home-assistant/home-assistant/compare/tibber_plot?expand=1#diff-43f211c43ec818fda11c8afe6bab1ee5R95

This will allow the Tibber sensor to show the coming electricity prices, and not only the historic ones.
I guess other components could use this functionality too. And I already have an idea of adding another Tibber sensor, where this will be required.

(I have not done much frontend development, so hope to get some good feedback)

### CODE REVIEW

**Comment by c727** [comment]
isn't it better to create a new domain for this?

**Comment by Danielhiversen** [comment]
> isn't it better to create a new domain for this?

Maybe.

But from a backend perspective it feels strange to change the domain to use a different frontend card. I think it would be strange to have the Tibber sensor as something else than a sensor.  We could also make it possible for the user to choose if (s)he wants a "normal" sensor where only the current state is shown or this new view with a graph. Then it would be strange if the domain changes according to this this. Then the entity_id will also change-

> I think we don't want to store tons of data in the state attributes so we will need an API for this

Maybe we should have an api/refactoring such that the frontend cards can be chosen independently of the domain. 


> Since we want to replace Google charts maybe you can check the PR #429 by @fanthos and use this lib

Yes, when this pr is merged, the plot_data cards also needs to be updated

**Comment by balloob** [comment]
This data should not be stored in the state machine. We should get it from an API. 

Cards are tied to a domain because of simplicity. If you want fancy stuff, use custom UI. It's otherwise unmaintainable for us as every card should be able to render every entity 🤔

**Comment by Danielhiversen** [comment]
i think this is such a generic feature, that it should be part of the official UI.

Ok, how can such an API be implemented? Do we have any examples?

**Comment by c727** [comment]
I'm not sure if it's a good example but I did something similar here https://github.com/home-assistant/home-assistant/pull/11866/files

**Comment by balloob** [comment]
I would say that adding / removing history entities is a great use case to use the new Config Entries component https://github.com/home-assistant/home-assistant/pull/12079 https://github.com/home-assistant/home-assistant-polymer/pull/861 ([this example shows all that we need to do](https://github.com/home-assistant/home-assistant/pull/12079/files#diff-809bbba84d6c4b392463cd9ac4b36c88))

**Comment by Danielhiversen** [comment]
> Would it make sense to allow the history component to make entities which show graphs for that entity ?

That could work. Then when have to accept future dates in the database. And the frontend should plot future dates if it exists.

**Comment by c727** [comment]
I think we should keep this generic graph card separated from the history graph to make it more flexible.
Someone posted this in the forums yesterday, maybe we can adapt some good stuff
https://community.home-assistant.io/t/google-charts-state-card-for-home-assistant/42400

**Comment by Danielhiversen** [comment]
Yes. To use the history graph would work for my use case.
But I think we should have a generic graph card, that allow a backend component to generate a custom graph.

**Comment by balloob** [comment]
We should not get a generic graph card. We always tie all frontend stuff to specific components. We can reuse building blocks that we use to build cards but won't have one card serve multiple domains.

If you want other data than history data you will  need a graph component, but how will platforms work for it? Maybe better for custom UI..

**Comment by Danielhiversen** [comment]
Ok, the history entities and using the history graph would work for my two use cases. But will we allow future dates to be added to database and visualized?

**Comment by Danielhiversen** [comment]
I have two use cases where I think this is required, and should be part of the official code. For the electricity price in Tibber we are currently showing the current price and the previous prices. But the future prices are much more interesting, since they can be used for automation and planing of the heating/ electricity usage. (Use less electricity when the price is high).

So I think we should support that sensor components can provide data, that are shown as a graph in the frontend.

I see two issues with my proposed code:

1) How to send the data to the frontend?
The data array for plotting is send to the frontend as an attribute. This mean the data array are stored in the database, and easily may take a lot disk space. This is the same way we do it for the weather component, so I thought it would be ok. The benefits of add the data as an attribute, is that you can easily use it for automation. 
One way to solve this, is to filter out all attributes starting with `_` from the database. Then we can rename `forecast` in the weather component to `_forecast`. This will also allow to add useful data to the state for automations, but without having to store it in the database.

2) How to choose UI card?
To day this is based on the domain name, but I think it should be possible for a sensor component to set some kind of a flag to use the `plot_data_card`
I have proposed one method, to set such a flag. But there are probably better ways.
The other option is to make a new domain (sensor_plot ?), that inherent the sensor component in the backend and adds a `plot_data` attribute.

(This is an important feature for me, so I really hope we can find a way to add it to the main repo.)

**Comment by balloob** [comment]
1. The recorder is going to record everything. It is not going to pick and choose attributes. It's opening a can of worms. We need to fetch that data via an API. We also can't store future dates in the recorder. it's only recording events/states that happened in the system.

2. This is exactly what the history entity would be, it would point at an entity and plot that graph.

**Comment by balloob** [comment]
Realized this already exists as [history_graph](https://home-assistant.io/components/history_graph/)

File: src/cards/ha-plot_data-card.html [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+          text-align: right;
+        }
+        span.line {
+          display: block;
+        }
+        .clear {
+          clear: both;
+        }
+    </style>
+    <google-legacy-loader on-api-load='googleApiLoaded'></google-legacy-loader>
```
Comment (andrey-git) [inline src/cards/ha-plot_data-card.html]: We have already switched to chartsjs

File: src/cards/ha-plot_data-card.html [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    <ha-card header='[[computeTitle(stateObj)]]'>
+      <div class='content'>
+          <div class='attribution attribution'>[[stateObj.state]] [[stateObj.attributes.unit_of_measurement]]</div>
```
Comment (andrey-git) [inline src/cards/ha-plot_data-card.html]: you have `attribution` twice

File: src/cards/ha-plot_data-card.html [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        <div id='chart_id' hidden$="[[!attr._plot_data]]"></div>
```
Comment (andrey-git) [inline src/cards/ha-plot_data-card.html]: attr is not defined anywhere

**Review by andrey-git — CHANGES_REQUESTED** [review]
I think such switching cards by _plot_data attribute is a hack.

**Comment by Danielhiversen** [comment]
@balloob 
No, the history_graph is not useful for me.
It plots historic data from the database. But I want to plot future values not in the database.
I have a similar use case as the weather component, where I want to plot the future electricity prices.

@andrey-git 
Thanks for your review. I know I have to update the code to use chartsjs.
> I think such switching cards by _plot_data attribute is a hack.

Yeah, you are probably right. But do you have any better solution?
I have two issues I have to solve for this:
1) How to send the data to the frontend?
2) How to choose UI card?

More details in https://github.com/home-assistant/home-assistant-polymer/pull/868#issuecomment-364374930

**Comment by magnusoverli** [comment]
Hi all!

I cannot contribute to the development of this, but I just wanted to let you know that the ability to pull future data on the electricity is a huge deal. It really is the difference between doing simple on/off automations and a more intelligent design.

Not sure I really understand the issues that have come up here, but couldn't´t this be solved by having another two sensor entity that always pulls data from 1 hour and 5 into the future?

**Comment by c727** [comment]
Let's close this PR. We no longer use Google charts and a solid implementation requires some more changes. Maybe you can discuss this in the architecture repo first

**Pull request closed without merging**